keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,870 @@
+know that this has succeeded?
+-->
+- Add a new field `maxRestartTimes` to the spec of pod which only applies when `restartPolicy` is `OnFailure`
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Combine the implementations of `backoffLimit` and `maxRestartTimes` together.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Add a new field `maxRestartTimes` to the spec of pod. `maxRestartTimes` only applies
+when `restartPolicy` is set to `OnFailure`. As a result, the restart times of a Pod
+will be controlled by `maxRestartTimes`.
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+When running with a healthy pod, some system errors or program internal errors
+encountered, which causes the application restarting infinitely. This is
+unrecoverable and wasteful.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+N/A
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+- Considering `Job` has also implemented the similar feature called `backoffLimit`,
+this can be confused to end-users how these two fields interacting with each other.
+- If `maxRestartTimes` configured in a recoverable circumstance with a small number,
+e.g. network shaking with 3 times, the service will stay in a failed state forever
+after 3 restarts. So end-users should be careful when configuring this field.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+If Pod's `RestartPolicy` is `Always` or `Never`, `MaxRestartTimes` is default to -1, and will not apply.
```

Comment:
We should make the field a `*int` - instead of using -1 value just use nil (which would also make it backward compatible).